### ISSUE TITLE
[Bug Report] enable async crashes or hangs in tt-train nano_gpt 

### ISSUE THREAD

**Issue by dmakoviichuk-tt** [tenstorrent/tt-metal#14171]
**Describe the bug**
If enable async in nano_gpt example.  it will crash or hang randomly.
Also performance in async mode is worse than without
**To Reproduce**
1. turn on async in Turn on caching in the https://github.com/tenstorrent/tt-train/blob/main/sources/examples/nano_gpt/main.cpp
2. run nano_gpt
3.  crash or hang

**Expected behavior**
It should work faster and don't hang.

**Additional context**
It is a big blocker for us. @davorchap  expects it will improve our performance.

**Comment by jvasilje** [comment]
@tt-asaigal any updates on this?

**Comment by tt-asaigal** [comment]
> @tt-asaigal any updates on this?

Hey @jvasilje: @davorchap , @dmakoviichuk-tt  and I spoke offline. The conclusions were as follows:
1. Async Mode isn't expected to help with single device use-cases. We see this with inference workloads in Metal as well: performance benefits are seen with multi-chip workloads.
2. Based on 1, I was told that the priority on this could be reduced for now since this isn't a blocker for single chip training.

If it makes sense for me to take a look now, I can do so.

**Comment by jvasilje** [comment]
@tt-asaigal @dmakoviichuk-tt re-labeling to P1.

**Comment by dmakoviichuk-tt** [comment]
@davorchap  @tt-asaigal  Since we already started our work on multiple devices in it nice have it done faster.

**Comment by prajaramanTT** [comment]
@dmakoviichuk-tt is this still an open issue ? If not, can you please close this ? Thanks.

**Comment by prajaramanTT** [comment]
@tt-asaigal / @davorchap Any updates on when this will be addressed ?

**Comment by tt-asaigal** [comment]
Hey, I ran this test with the following line added back to enable async mode:
`device->enable_async(true);`

I have run this test multiple times, for an hour at this point and I don't see an issue. The loss value converges with or without async enabled.

Performance improvements on single device workloads are not expected, since the main thread blocks on outputs from the worker thread quite frequently.

Please let me know if you're able to reproduce successful runs on your end or if I'm missing anything.

**Comment by dmakoviichuk-tt** [comment]
hi @tt-asaigal  we didn't test it for a couple weeks. Latest crash fixes (matmuls and unary ops) might have fix it too.
Let me doublecheck it. Thank you for your feedback.

**Comment by dmakoviichuk-tt** [comment]
@tt-asaigal just ran it in release nanogpt.
```
Full step time 288.005 ms, cache entries: 96
dataloader host only step time 0.241 ms
dataloader step time 6.236 ms
Step: 101, Loss: 2.546875
Full step time 287.316 ms, cache entries: 96
dataloader host only step time 0.045 ms
dataloader step time 6.184 ms
Step: 102, Loss: 2.53125
Full step time 5488.304 ms, cache entries: 96
dataloader host only step time 0.048 ms
dataloader step time 5.648 ms
```

If you see last step takes 20x time to run and the it hangs.

If I disable async everything is fine.

**Comment by dmakoviichuk-tt** [comment]
Building the latest main now. I tested on yesterdays version.

**Comment by tt-asaigal** [comment]
What machine are you running on? I used an N300 and was not able to repro. Apart from enabling async, was there any other change?

**Comment by dmakoviichuk-tt** [comment]
@tt-asaigal  unfortunately on n150 I see frequent freezes when run training even in sync mode.

**Comment by dmakoviichuk-tt** [comment]
@bbradelTT seems like its matmul.
I was able to see hung in the debugger and it happend in the matmul op.
To reproduce:
Run nanogpt couple of times. Last time it crashed ~ 1-2 hours of runs. But it is very often for me.
Also didn't happen in n300.

**Comment by dmakoviichuk-tt** [comment]
Moved to the P0 because it is matmul.
cc @davorchap

**Comment by davorchap** [comment]
Does this require async bug fix and perf improvements?

**Comment by bbradelTT** [comment]
@dmakoviichuk-tt can this issue be reproduced in a smaller amount of time, e.g. by lowering the max steps or creating a minimal reproducible example?

Having to wait 1-2 hours before the issue appears is not going to work, both because it would take too long to iterate on solutions and the amount of debug/dprint data generated would be too large.

**Comment by bbradelTT** [comment]
Couldn't reproduce after running for around 3 hours on an N150.

Trying with gpt2s config: `build/sources/examples/nano_gpt/nano_gpt -c /localdev/bbradel/tt-metal/tt-train/configs/training_shakespear_gpt2s.yaml`

**Comment by bbradelTT** [comment]
I will also try the following with and without watcher (with a larger value of n). @dmakoviichuk-tt said that is the likely shape that is causing the problem.
```
import pytest
import torch
import ttnn
from tests.ttnn.utils_for_testing import assert_with_pcc

def test_mm(device):
    x = torch.randn((1, 1, 64 * 256, 384 * 4), dtype=torch.float32)
    y = torch.randn((1, 1, 4 * 384, 384), dtype=torch.float32)
    x_tt = ttnn.from_torch(x, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
    y_tt = ttnn.from_torch(y, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
    n = 5
    for i in range(n):
        z_tt = ttnn.matmul(x_tt, y_tt)
    z = ttnn.to_torch(z_tt)
    z_t = torch.matmul(x, y)
    print(f'z_t {z_t}\nz_tt {z_tt}\nz {z}')
    assert_with_pcc(z_t, z, pcc=0.999)
```

**Comment by bbradelTT** [comment]
To see progress, following code could be added where the for loop is:
```
    m = 10000
    for i in range(n):
        if i % m == 0:
            print(f"i {i} / {n} : {100.0 * i/n}%")
        z_tt = ttnn.matmul(x_tt, y_tt)
```

Tried with n=500000 and watcher off. So far no luck.

**Comment by bbradelTT** [comment]
This is going to be slow moving, at least for a while. Moving to a P1 until we get a good repro.

**Comment by dmakoviichuk-tt** [comment]
@bbradelTT seems like it is not a matmul.
@tt-asaigal @cfjchu we have 100% reproduce with async true. 
It became very important to us. I am returning back p0.
@rfurko-tt will provide test branch and more details.

**Comment by rfurko-tt** [comment]
Branch: `rfurko/showcase_async_issue_with_loudbox`

Instructions to reproduce:
* Use t3k device (4xN300)
* Build tt-metal with tt-train: `./build_metal.sh -b Release --build-tt-train`
* Run command: `TT_METAL_LOGGER_LEVEL=FATAL ./build/tt-train/sources/examples/nano_gpt/nano_gpt -w 0 --ddp 1 -c tt-train/configs/training_shakespear_nanogpt.yaml`

Out of 5 runs: 
* 1 freeze
* 4 crashes (<100-500 steps, it will be really fast, each step takes ~60ms)

**Comment by cfjchu** [comment]
I ran your repro and it completely killed my machine 😅 . It did print this out before it died:

```
libc++abi: terminating due to uncaught exception of type std::runtime_error: TT_THROW @ /localdev/jchu/tt-metal/workspace0/ttnn/cpp/ttnn/operations/matmul/matmul.cpp:68: tt::exception
info:
ttnn.matmul: The width of the first tensor must be equal to the height of the second tensor (96 != 2048). The shape of first tensor was Shape([2048, 96]) and the shape of second tensor was Shape([2048, 384]))
backtrace:
 --- /localdev/jchu/tt-metal/workspace0/build_Release/ttnn/_ttnn.so(+0x2084ae5) [0x7fef12886ae5]
 --- ttnn::operations::matmul::bound_matmul(tt::tt_metal::Tensor const&, tt::tt_metal::Tensor const&, std::__1::optional<tt::tt_metal::Tensor const> const&, ttnn::operations::matmul::Matmul const&, unsigned char const&, std::__1::optional<tt::tt_metal::Tensor>&)
 --- /localdev/jchu/tt-metal/workspace0/build_Release/ttnn/_ttnn.so(_ZN4ttnn10operations6matmul15MatmulOperation6invokeERKN2tt8tt_metal6TensorES7_bbRKNSt3__18optionalIKNS4_12MemoryConfigEEENS9_IKNS4_8DataTypeEEERKNS9_IKNS8_7variantIJNS1_28MatmulMultiCoreProgramConfigENS1_45MatmulMultiCoreNonOptimizedReuseProgramConfigENS1_33MatmulMultiCoreReuseProgramConfigENS1_42MatmulMultiCoreReuseMultiCastProgramConfigENS1_44MatmulMultiCoreReuseMultiCast1DProgramConfigENS1_53MatmulMultiCoreReuseMultiCastDRAMShardedProgramConfigEEEEEERKNS9_IKNS8_12basic_stringIcNS8_11char_traitsIcEENS8_9allocatorIcEEEEEENS9_IKNSI_IJNS_28GrayskullComputeKernelConfigENS_27WormholeComputeKernelConfigEEEEEENS9_IKNS_5types8CoreGridEEERKNS9_IKNS4_4TileEEENS9_IS5_EERKNS9_IKNSI_IJNS8_9monostateENS4_12experimental20GlobalCircularBufferENS_22global_circular_buffer31MultiDeviceGlobalCircularBufferEEEEEERKNS9_INS3_3stl10StrongTypeIhNS4_14SubDeviceIdTagEEEEE+0x367) [0x7fef12886e57]
 --- ./build/tt-train/sources/examples/nano_gpt/nano_gpt(+0x440fa0) [0x563ee1ac9fa0]
 --- ./build/tt-train/sources/examples/nano_gpt/nano_gpt(+0x440c70) [0x563ee1ac9c70]
 --- ./build/tt-train/sources/examples/nano_gpt/nano_gpt(+0x43fc5b) [0x563ee1ac8c5b]
 --- ./build/tt-train/sources/examples/nano_gpt/nano_gpt(+0x3e02a2) [0x563ee1a692a2]
 --- ./build/tt-train/sources/examples/nano_gpt/nano_gpt(+0x3e1431) [0x563ee1a6a431]
 --- ./build/tt-train/sources/examples/nano_gpt/nano_gpt(+0x36f169) [0x563ee19f8169]
 --- ./build/tt-train/sources/examples/nano_gpt/nano_gpt(+0x2f9b1a) [0x563ee1982b1a]
 --- /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7fef0f59cd90]
 --- /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80) [0x7fef0f59ce40]
 --- ./build/tt-train/sources/examples/nano_gpt/nano_gpt(+0x2f2ed5) [0x563ee197bed5]
```

I just wanted to see if it was something straightforward to fix, but it does not seem to be the case. Can this wait 1-2 more weeks? We'll be ditching this multi-threaded async dispatch onto device so there's no point in sinking resources to fix this.

**Comment by Denys88** [comment]
@cfjchu seems like it is caused by using  sync version of the shape() somewhere inside of the ttnn code, I am pretty sure we fixed this calls in our codebase some time ago.
So I believe there is a very straightforward solution for this problem just need to find this shape.
Overall it reveals some design level issues for our api. (Myabe need to remove calls without locks at all?)
if you can provide a quick fix there is a chance I can demonstrate a loudbox training next week during tt-days.

**Comment by cfjchu** [comment]
Let me try and run some experiments in the evening to see whether we can pinpoint the issue. Any additional information or smaller repro you can share would be helpful.

**Comment by cfjchu** [comment]
Just to follow-up here as well. I looked into this last night and shared the update with @rfurko-tt and @dmakoviichuk-tt offline.

I confirmed the bug seems to be related to some problems with async-mode enabled. After trying some changes with multiple experiments, I wasn't able to pin-point the issue.

However, now that we have CCLs working on our TT-Mesh integration branch, I was able to run this test successfully there!! Relative to `main`, it's faster than sync-mode and training converges. 

![Image](https://github.com/user-attachments/assets/c095d2c7-3eeb-4574-9897-d9c6d7ca52bf)

I'll downgrade this issue to P2 since it works on our integration branch and we won't be spending time/resources to resolve the original problem.
